--- src/theory/theory_engine.cpp.orig	2013-02-13 10:18:48.000000000 -0500
+++ src/theory/theory_engine.cpp	2013-02-13 10:19:06.000000000 -0500
@@ -826,7 +826,8 @@
 
     // If this is an atom, we preprocess its terms with the theory ppRewriter
     if (Theory::theoryOf(current) != THEORY_BOOL) {
-      d_ppCache[current] = ppTheoryRewrite(current);
+      Node ppRewritten = ppTheoryRewrite(current);
+      d_ppCache[current] = ppRewritten;
       Assert(Rewriter::rewrite(d_ppCache[current]) == d_ppCache[current]);
       continue;
     }
