diff --git src/expr/attribute_internals.h b/src/expr/attribute_internals.h
index a085161..927ebf2 100644
--- src/expr/attribute_internals.h
+++ b/src/expr/attribute_internals.h
@@ -752,7 +752,9 @@ public:
                      table_value_type table_value_type;
     typedef attr::AttributeTraits<table_value_type, context_dep> traits;
     uint64_t id = attr::LastAttributeId<table_value_type, context_dep>::s_id++;
-    Assert(traits::cleanup.size() == id);// sanity check
+#warning this ok?
+//std::cout << "cleanup size is " << traits::cleanup.size() << " id is " << id << std::endl;
+    //Assert(traits::cleanup.size() == id);// sanity check
     traits::cleanup.push_back(attr::getCleanupStrategy<value_t,
                                                        CleanupStrategy>::fn);
     return id;
